experimental/oak_async/src/grpc.rs

```diff
@@ -0,0 +1,146 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! gRPC related types for asynchronous processing.
+
+use log::error;
+use oak::{
+    grpc::{ChannelResponseWriter, Code, WriteMode},
+    OakError,
+};
+
+/// Writes exactly one message on a gRPC response channel.
+///
+/// This type is used for gRPC methods where the return type is non-streaming. Instances must be
+/// consumed using `send`. In exceptional circumstances the `into_inner` method may be used to
+/// access the underlying `ChannelResponseWriter`.
+///
+/// If neither `send` nor `into_inner` is called, an `UNKNOWN` error will be written as the response
+/// upon `Drop`.
+#[must_use = "exactly one message should be sent. Use into_inner for advanced use cases"]
+pub struct OneshotWriter<T> {
+    writer: ChannelResponseWriter,
+    _msg: core::marker::PhantomData<T>,
+}
+
+impl<T: ::prost::Message> OneshotWriter<T> {
+    // Only for use by the Oak Service Generator.
+    #[doc(hidden)]
+    pub fn new(writer: ChannelResponseWriter) -> OneshotWriter<T> {
+        OneshotWriter {
+            writer,
+            _msg: core::marker::PhantomData,
+        }
+    }
+
+    /// Send the response, automatically closing the channel.
+    pub fn send(self, msg: &T) -> Result<(), OakError> {
+        self.writer.write(msg, WriteMode::Close)?;
+        // Prevents running the destructor
+        core::mem::forget(self);
+        Ok(())
+    }
+
+    /// Consumes this writer, returning a `ChannelResponseWriter` in its place.
+    pub fn into_inner(self) -> ChannelResponseWriter {
+        into_inner(self, |s| &mut s.writer)
+    }
+}
+
+impl<T> Drop for OneshotWriter<T> {
+    /// Close the channel with an `Unknown` status.
+    fn drop(&mut self) {
+        // TODO(wildarch) or should we panic (after setting the error)?
+        error!("drop() called on OneshotWriter");
+        if let Err(e) = self
+            .writer
+            .close(Err(oak::grpc::build_status(Code::Unknown, "")))
+        {
+            error!("Failed to close OneshotWriter in Drop: {}", e);
+        };
+    }
+}
+
+/// Writes multiple messages to a gRPC response channel.
+///
+/// This type is used for gRPC methods where the return type is streaming.
+///
+///
+/// In exceptional circumstances the `into_inner` method may be used to access the underlying
+/// `ChannelResponseWriter`.
+pub struct MultiWriter<T> {
+    writer: ChannelResponseWriter,
```

Comment:
As above, I think this can be shorter and clearer with an `Option<ChannelResponseWriter>` here.

Reply:
Thanks, that's a good idea!

Done.